org.osgi.test.cases.jdbc/src/org/osgi/test/cases/jdbc/junit/JDBCTestCase.java

```diff
@@ -118,66 +129,57 @@ public void testRegisteredProperties(
-
-		assertThat(dsf.createDataSource(new DfltProps())).isNotNull();
+	@ServiceSource(serviceType = DataSourceFactory.class)
+	public void testCapabilityDataSource(ServiceReference<DataSourceFactory> sr)
+			throws SQLException, InvalidSyntaxException {
+		testCapability(sr, FILTER_CAPABILITY_DATASOURCE, DataSourceFactory::createDataSource);
 	}
 
 	@ParameterizedTest
-	@ServiceSource(serviceType = DataSourceFactory.class, filter = FILTER_CAPABILITY_DRIVER)
-	public void testCapabilityDriver(DataSourceFactory dsf)
-			throws SQLException {
-
-		assertThatNoException().isThrownBy(() -> dsf.createDriver(null));
-
-		assertThat(dsf.createDriver(null)).isNotNull();
-
-		assertThatNoException()
-				.isThrownBy(() -> dsf.createDriver(new Properties()));
-
-		assertThat(dsf.createDriver(new Properties())).isNotNull();
-
+	@ServiceSource(serviceType = DataSourceFactory.class)
+	public void testCapabilityDriver(ServiceReference<DataSourceFactory> sr)
+			throws SQLException, InvalidSyntaxException {
+		testCapability(sr, FILTER_CAPABILITY_DRIVER, DataSourceFactory::createDriver);
 	}
 
 	@ParameterizedTest
-	@ServiceSource(serviceType = DataSourceFactory.class, filter = FILTER_CAPABILITY_XADATASOURCE)
-	public void testCapabilityXADataSource(DataSourceFactory dsf)
-			throws SQLException {
-
-		assertThatNoException().isThrownBy(() -> dsf.createXADataSource(null));
-
-		assertThat(dsf.createXADataSource(null)).isNotNull();
-
-		assertThatNoException()
-				.isThrownBy(() -> dsf.createXADataSource(new DfltProps()));
-
-		assertThat(dsf.createXADataSource(new DfltProps())).isNotNull();
+	@ServiceSource(serviceType = DataSourceFactory.class)
+	public void testCapabilityXADataSource(ServiceReference<DataSourceFactory> sr)
+			throws SQLException, InvalidSyntaxException {
+		testCapability(sr, FILTER_CAPABILITY_XADATASOURCE, DataSourceFactory::createXADataSource);
+	}
+	
+	private <T> void testCapability(ServiceReference<DataSourceFactory> reference, String filter,
+			DsfCallback<DataSourceFactory, Properties, T> function) throws SQLException, InvalidSyntaxException {
+		DataSourceFactory dsf = bundleContext.getService(reference);
+		try {
+			if (FrameworkUtil.createFilter(filter).match(reference)) {
+				//if the source declares the support, it should work to create items without an exception
+				assertThatNoException().isThrownBy(() -> function.invoke(dsf, null));
+				assertThat(function.invoke(dsf, null)).isNotNull();
+				assertThatNoException().isThrownBy(() -> function.invoke(dsf, new DfltProps()));
+				assertThat(function.invoke(dsf, new DfltProps())).isNotNull();
+			} else {
+				//it is required to throw an exception here
+				assertThatExceptionOfType(SQLException.class).isThrownBy(() -> function.invoke(dsf, null));
+				assertThatExceptionOfType(SQLException.class).isThrownBy(() -> function.invoke(dsf, new DfltProps()));
+			}
+		} finally {
+			if (dsf != null) {
+				bundleContext.ungetService(reference);
```

Comment:
Not sure if this is needes. The injected BundleContext is [CloseableBundleContext](https://github.com/osgi/osgi-test/blob/main/org.osgi.test.common/src/main/java/org/osgi/test/common/context/CloseableBundleContext.java).

Reply:
Hard to guess, there is not a single line of documentation at that class, but it seems not wrong to be nice and cleanup ourselfs :-)